### PR TITLE
Support InstantX/InstantID

### PR DESCRIPTION
**Pull request by alekseyfa**
# What does this PR do?

This PR adds support for InstantX/InstantID.
Link to the original repository: https://github.com/InstantID/InstantID

Some of the code was copied from the original repository:
- The entire ip_adapter
-  GaudiStableDiffusionXLInstantIDPipeline is based on StableDiffusionXLInstantIDPipeline

A brief summary of the changes:
- Added GaudiStableDiffusionXLInstantIDPipeline inherited from GaudiDiffusionPipeline and StableDiffusionXLControlNetPipeline for building Stable Diffusion XL pipelines
- text_to_image_generation.py script was modified according to the existing logic to use InstantID pipeline. If a control image and Stable Diffusion XL model is specified, the InstandID functionality will be used

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

### CODE REVIEW

**Comment by alekseyfa** [comment]
Hi @regisss, could you please leave you review in this PR? Thanks!

File: optimum/habana/diffusers/pipelines/instantid/pipeline_stable_diffusion_xl_instantid.py [hunk 1]
```diff
@@ -0,0 +1,1137 @@
...
+                prompt_embeds_batch = prompt_embeds_batches[0]
+                prompt_embeds_batches = torch.roll(prompt_embeds_batches, shifts=-1, dims=0)
+                add_text_embeddings_batch = add_text_embeddings_batches[0]
+                add_text_embeddings_batches = torch.roll(add_text_embeddings_batches, shifts=-1, dims=0)
+                add_time_ids_batch = add_time_ids_batches[0]
+                add_time_ids_batches = torch.roll(add_time_ids_batches, shifts=-1, dims=0)
+                encoder_hidden_states = torch.cat([prompt_embeds_batch, prompt_image_emb], dim=1)
+                num_warmup_steps = len(timesteps) - num_inference_steps * self.scheduler.order
+
+                for i in range(num_inference_steps):
```
Comment (dsocek) [inline optimum/habana/diffusers/pipelines/instantid/pipeline_stable_diffusion_xl_instantid.py:781]: Please change `range(num_inference_steps)`  -> `range(len(timesteps))`

to cover some schedulers which could give num of timesteps != num_inference_steps

Same fix as here:
https://github.com/huggingface/optimum-habana/blob/main/optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L515

File: examples/stable-diffusion/text_to_image_generation.py [hunk 2]
```diff
@@ -295,32 +275,82 @@ def main():
...
-            guidance_scale=args.guidance_scale,
-            negative_prompt=args.negative_prompts,
-            eta=args.eta,
-            output_type=args.output_type,
-            profiling_warmup_steps=args.profiling_warmup_steps,
-            profiling_steps=args.profiling_steps,
-            **res,
-        )
+        if sdxl:
+            from optimum.habana.diffusers import GaudiStableDiffusionXLInstantIDPipeline
```
Comment (dsocek) [inline examples/stable-diffusion/text_to_image_generation.py:284]: Hmm we should not hijack this case strictly for InstantID sample.   The standard `GaudiStableDiffusionXLControlNetPipeline` should be default here, so please adjust logic.   See example here:
https://github.com/dsocek/optimum-habana/commit/64e56de700202ffb254c0d033fe7a7637b9a8211

Basically need something like:
```
if sdxl and controlnet:
  if instantid:
    # InstantID pipe 
 else:
    # default SDXL ControlNet pipe
```

File: optimum/habana/diffusers/pipelines/instantid/pipeline_stable_diffusion_xl_instantid.py [hunk 3]
```diff
@@ -0,0 +1,1137 @@
...
+
+            hb_profiler = HabanaProfile(
+                warmup=profiling_warmup_steps,
+                active=profiling_steps,
+                record_shapes=False,
+            )
+            hb_profiler.start()
+
+            # 8. Denoising loop
+            throughput_warmup_steps = kwargs.get("throughput_warmup_steps", 3)
```
Comment (dsocek) [inline optimum/habana/diffusers/pipelines/instantid/pipeline_stable_diffusion_xl_instantid.py:763]: Please use updated logic for warmup here.  See https://github.com/huggingface/optimum-habana/blob/main/optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L498
https://github.com/huggingface/optimum-habana/blob/main/optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L507
https://github.com/huggingface/optimum-habana/blob/main/optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L601

**Review by dsocek — CHANGES_REQUESTED** [review]
Looks good, but there are few fixes needed (see code comments).   Also, please check with @libin if we should have a separate model download script or integrated inside single sample script.

**Comment by emascarenhas** [comment]
@alekseyfa , Can you also provide performance numbers.

File: optimum/habana/diffusers/pipelines/instantid/ip_adapter/attention_processor.py
Comment (skavulya) [inline optimum/habana/diffusers/pipelines/instantid/ip_adapter/attention_processor.py:1]: Are there any gaudi-specific changes to the files in `ip_adapter`? If not, it would be cleaner to pip install instantid from git and import the required functions

File: optimum/habana/diffusers/pipelines/instantid/pipeline_stable_diffusion_xl_instantid.py [hunk 4]
```diff
@@ -0,0 +1,1137 @@
...
+from ....utils import HabanaProfile, speed_metrics
+from ..pipeline_utils import GaudiDiffusionPipeline
+from ..stable_diffusion_xl.pipeline_stable_diffusion_xl import GaudiStableDiffusionXLPipeline, GaudiStableDiffusionXLPipelineOutput
+from ..stable_diffusion.pipeline_stable_diffusion import retrieve_timesteps
+
+
+logger = logging.get_logger(__name__)
+
+
+class GaudiStableDiffusionXLInstantIDPipeline(GaudiDiffusionPipeline, StableDiffusionXLControlNetPipeline):
```
Comment (skavulya) [inline optimum/habana/diffusers/pipelines/instantid/pipeline_stable_diffusion_xl_instantid.py:66]: Please extend `StableDiffusionXLInstantIDPipeline` and only override the methods that are hpu-specific

File: examples/stable-diffusion/text_to_image_generation.py (hunk 2 above)
Comment (alekseyfa) [reply examples/stable-diffusion/text_to_image_generation.py:311]: I've added `--ip_adapter_path` argument that allows to delineate the logic. If this argument is specified, `GaudiStableDiffusionXLInstantIDPipeline` is used

File: optimum/habana/diffusers/pipelines/instantid/pipeline_stable_diffusion_xl_instantid.py (hunk 1 above)
Comment (alekseyfa) [reply optimum/habana/diffusers/pipelines/instantid/pipeline_stable_diffusion_xl_instantid.py]: Thanks! done

File: optimum/habana/diffusers/pipelines/instantid/pipeline_stable_diffusion_xl_instantid.py (hunk 3 above)
Comment (alekseyfa) [reply optimum/habana/diffusers/pipelines/instantid/pipeline_stable_diffusion_xl_instantid.py:767]: updated

File: optimum/habana/diffusers/pipelines/instantid/ip_adapter/attention_processor.py
Comment (alekseyfa) [reply optimum/habana/diffusers/pipelines/instantid/ip_adapter/attention_processor.py:1]: There are no Gaudi specific changes in `ip_adapter`, it was just copied. Unfortunately the installation of Instant ID is not possible as the guys do not provide any installation files or scripts

File: optimum/habana/diffusers/pipelines/instantid/pipeline_stable_diffusion_xl_instantid.py (hunk 4 above)
Comment (alekseyfa) [reply optimum/habana/diffusers/pipelines/instantid/pipeline_stable_diffusion_xl_instantid.py:66]: Due to the impossibility of installing Instant ID for inheritance from [StableDiffusionXLInstantIDPipeline](https://github.com/InstantID/InstantID/blob/main/pipeline_stable_diffusion_xl_instantid.py#L136), it was decided to simply adapt the class to Gaudi by copying some of the methods and functions

**Comment by alekseyfa** [comment]
> @alekseyfa , Can you also provide performance numbers.

The results so far on single HPU are as follows

```bash
python text_to_image_generation.py \
    --model_name_or_path stabilityai/stable-diffusion-xl-base-1.0 \
    --controlnet_model_name_or_path checkpoints/ControlNetModel \
    --prompts "analog film photo of a woman. faded film, desaturated, 35mm photo, grainy, vignette, vintage, Kodachrome, Lomography, stained, highly detailed, found footage, masterpiece, best quality" \
    --negative_prompts "(lowres, low quality, worst quality:1.2), (text:1.2), watermark, painting, drawing, illustration, glitch, deformed, mutated, cross-eyed, ugly, disfigured (lowres, low quality, worst quality:1.2), (text:1.2), watermark, painting, drawing, illustration, glitch,deformed, mutated, cross-eyed, ugly, disfigured" \
    --control_image https://hf.co/datasets/huggingface/documentation-images/resolve/main/diffusers/input_image_vermeer.png \
    --ip_adapter_path checkpoints/ip-adapter.bin \
    --control_preprocessing_type "none" \
    --image_save_dir /tmp/stable_diffusion_images \
    --use_habana \
    --use_hpu_graphs \
    --gaudi_config Habana/stable-diffusion \
    --num_images_per_prompt 10 \
    --batch_size 8 \
    --bf16
```

`Speed metrics: {'generation_runtime': 336.1959, 'generation_samples_per_second': 0.846, 'generation_steps_per_second': 0.106}`

To compare the results on ControlNet single HPU:

```bash
python text_to_image_generation.py \
    --model_name_or_path runwayml/stable-diffusion-v1-5 \
    --controlnet_model_name_or_path lllyasviel/sd-controlnet-openpose \
    --prompts "analog film photo of a woman. faded film, desaturated, 35mm photo, grainy, vignette, vintage, Kodachrome, Lomography, stained, highly detailed, found footage, masterpiece, best quality" \
    --control_image https://hf.co/datasets/huggingface/documentation-images/resolve/main/diffusers/input_image_vermeer.png \
    --control_preprocessing_type "none" \
    --num_images_per_prompt 10 \
    --batch_size 8 \
    --image_save_dir /tmp/controlnet_images \
    --use_habana \
    --use_hpu_graphs \
    --gaudi_config Habana/stable-diffusion \
    --bf16
```

`Speed metrics: {'generation_runtime': 133.4811, 'generation_samples_per_second': 0.12, 'generation_steps_per_second': 0.015}`

And Stable Diffusion XL (SDXL) single HPU:

```bash
python text_to_image_generation.py \
    --model_name_or_path stabilityai/stable-diffusion-xl-base-1.0 \
    --prompts "Sailing ship painting by Van Gogh" \
    --num_images_per_prompt 20 \
    --batch_size 4 \
    --image_save_dir /tmp/stable_diffusion_xl_images \
    --scheduler euler_discrete \
    --use_habana \
    --use_hpu_graphs \
    --gaudi_config Habana/stable-diffusion \
    --bf16
```

`{'generation_runtime': 314.1854, 'generation_samples_per_second': 0.225, 'generation_steps_per_second': 0.14}`

File: optimum/habana/diffusers/pipelines/instantid/pipeline_stable_diffusion_xl_instantid.py (hunk 4 above)
Comment (skavulya) [reply optimum/habana/diffusers/pipelines/instantid/pipeline_stable_diffusion_xl_instantid.py:66]: Sounds good.

**Review by dsocek — APPROVED** [review]
LGTM, please rebase

File: examples/stable-diffusion/text_to_image_generation.py (hunk 2 above)
Comment (alekseyfa) [reply examples/stable-diffusion/text_to_image_generation.py:363]: done

File: tests/test_diffusers.py [hunk 5]
```diff
@@ -5074,3 +5075,270 @@ def test_stable_diffusion_xl_inpaint_no_throughput_regression(self):
...
+            "controlnet_conditioning_scale": 0.8,
+            "ip_adapter_scale": 0.8,
+            "generator": generator,
+            "output_type": "np",
+            "image": image,
+        }
+        return inputs
+
+    def test_stable_diffusion_xl_instantid_default(self):
+        # device = "cpu"  # ensure determinism for the device-dependent torch.Generator
```
Comment (imangohari1) [inline tests/test_diffusers.py:5160]: remove this if not needed.

**Review by imangohari1 — COMMENTED** [review]
I have not looked at this PR carefuly, but please run the CI jobs as below and share the results. 
They should pass 
```bash 
python -m pip install .[tests]
python -m pytest tests/test_diffusers.py -v -s -k test_stable_diffusion_xl_instantid
```

**Comment by alekseyfa** [comment]
> I have not looked at this PR carefuly, but please run the CI jobs as below and share the results. They should pass
> 
> ```shell
> python -m pip install .[tests]
> python -m pytest tests/test_diffusers.py -v -s -k test_stable_diffusion_xl_instantid
> ```

Thanks for your help! I've updated the tests, now everything passes

File: tests/test_diffusers.py (hunk 5 above)
Comment (alekseyfa) [reply tests/test_diffusers.py]: fixed
